Describe the behavior of different token formats in the HTTP JSON API

stefanobaghino-da opened the pull request:
Fixes #12440

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

realvictorprm commented on docs/source/json-api/index.rst:
```diff
@@ -266,12 +266,16 @@ The JSON API essentially performs two separate tasks:
+Party-specific requests, i.e., command submissions and queries, require a JWT that demonstrates the bearer's authorization to issue the request.
 
-.. note:: While the JSON API receives the token it doesn't validate it itself. Upon receiving a token it will pass it, and all data contained within the request, on to the Ledger API's AuthService which will then determine if the token is valid and authorized. However, the JSON API does decode the token to extract the ledger id, application id and party so it requires that you use the JWT format  ...
+If the underlying ledger supports :ref:`user management <user-service>` (this includes Canton and the sandbox), the only restriction is imposed by the authorization: the bearer's user must have the rights to perform the operation. The HTTP JSON API will parse the token, read the user id and fill it into the Ledger API request's application id field.
```
not certain about the part
> read the user id and fill it into the Ledger API request's application id field.

We do read the user id from the token but we only do so in endpoints that require the user id for calls which so far are only a few user management endpoints (i.e. /user to get user information about the user associated with the token)

cocreature replied:
See https://github.com/digital-asset/daml/blob/728f03c85546ba88111114a8b3b17fc91aaa9adb/ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala#L159

You also use that on command submissions to fill in the application id as far as I know.

stefanobaghino-da replied:
@realvictorprm Is @cocreature's point valid or not? You wrote the code, I assume you're in the best position to tell whether I'm writing something correct or not.

realvictorprm replied:
@cocreature definitely is right (as the code shows) and with me additionally remembering that this is the case because we need to provide an application id such that we decided on to use the user-id for that (all makes sense again to me xD). Very sorry for the confusion.

stefanobaghino-da replied:
No problem, thanks for the clarification.

realvictorprm commented on docs/source/json-api/index.rst:
```diff
@@ -266,12 +266,16 @@ The JSON API essentially performs two separate tasks:
+-----------------------
 
-Party-specific requests, i.e., command submissions and queries, require a JWT with some additional restrictions compared to general :ref:`Daml ledger access token formats <access-token-formats>`. For command submissions, ``actAs`` must contain at least one party and ``readAs`` can contain 0 or more parties. Queries require at least one party in either ``actAs`` or ``readAs``. The application id a ...
+Party-specific requests, i.e., command submissions and queries, require a JWT that demonstrates the bearer's authorization to issue the request.
```
This paragraph is not very informative to me, what is a JWT that demonstrates the bearer's authorization?

And should we still include it like this if the user management shall be the recommended way of dealing with tokens? I.e. shouldn't the user management way be mentioned first and then mentioned as legacy approach the party specific tokens?

PS: Documentation is always very fuzzy thing to solve, these are just my thoughts on this.

stefanobaghino-da replied:
Is your ask to remove this paragraph altogether or something else? I cannot tell.

realvictorprm reviewed (commented):
Unfortunately I'm not fully bought into the changes here and I think it would be more informative and helpful to the user to keep the paragraph about the `Party-specific Access Tokens` while providing another paragraph on `User Access Tokens` which should appear first with explicitly mentioning that these are the way to go from now on and that the other format is deprecated.

However I do think the changes were made at the right position, it definitely lacks information at that position.

stefanobaghino-da commented:
> Unfortunately I'm not fully bought into the changes here and I think it would be more informative and helpful to the user to keep the paragraph about the `Party-specific Access Tokens` while providing another paragraph on `User Access Tokens` which should appear first with explicitly mentioning that these are the way to go from now on and that the other format is deprecated.
> 
> However I do think the changes were made at the right position, it definitely lacks information at that position.

Cool, I'm happy to make the changes that you think are necessary. What would those be?

stefanobaghino-da commented:
> Unfortunately I'm not fully bought into the changes here and I think it would be more informative and helpful to the user to keep the paragraph about the `Party-specific Access Tokens` while providing another paragraph on `User Access Tokens` which should appear first with explicitly mentioning that these are the way to go from now on and that the other format is deprecated.
> 
> However I do think the changes were made at the right position, it definitely lacks information at that position.

I'm not sure of what you mean by "keep the paragraph about the `Party-specific Access Tokens`". That paragraph was not about custom claims token, but rather about issuing requests that required party information. I changed the tail to avoid confusion. The information is still there, I moved the content about custom claims token down in a note after the paragraph about user management. I don't think we're fully ready to document that custom claim tokens are deprecated.

realvictorprm commented:
> I don't think we're fully ready to document that custom claim tokens are deprecated.

I wasn't aware of that, then let revisit the documentation with more specific actions (I haven't added much myself because I wasn't overall sure yet and wanted to see first what the others think about the changes).

realvictorprm commented on docs/source/json-api/index.rst:
```diff
@@ -266,12 +266,34 @@ The JSON API essentially performs two separate tasks:
+
+If the underlying ledger supports :ref:`user management <user-service>` (this includes Canton and the sandbox), you are
+recommended to use user tokens. For command submissions, the user id of the bearer should have ``actAs`` rights for at
+least one party and ``readAs`` rights for any number of parties. Queries require the bearer's user id to have at least
```
I think `user` fits better here than `user id` (same for the line above this one).

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/commit/1fcb20ff7c1ca1c19932a8c64d17193f13e30807

realvictorprm commented on docs/source/json-api/index.rst:
```diff
@@ -266,12 +266,34 @@ The JSON API essentially performs two separate tasks:
+^^^^^^^^^^^^^^^^^^
+
+These tokens can be used if the underlying ledger does not support :ref:`user management <user-service>`. For command
+submissions, ``actAs`` must contain at least one party and ``readAs`` can contain 0 or more parties. Queries require at
```
I would align this with the description above i.e. `can have any number of parties` instead of `can contain 0 or more parties`.

realvictorprm reviewed (approved):
This is great, thanks!

S11001001 commented on docs/source/json-api/index.rst:
```diff
@@ -266,12 +266,36 @@ The JSON API essentially performs two separate tasks:
 
-Party-specific Access Tokens
-----------------------------
+Party-specific Requests
```
Not a big fan of this header, but I don't have any suggestions, and looks good otherwise.

stefanobaghino-da replied:
It seemed like a good fit for the paragraph (since it's the subject in the very first sentence) and I believe it lowers the possibility of somehow confusing this paragraph to something more closely related to something specific to custom claims access tokens. If by any chance you get a better idea, I'm open to making changes of course.
